Internal components refactor [g3w-client #574]

Raruto opened the pull request:
## List of changes

1. internal JS/Vue components refactor: https://github.com/g3w-suite/g3w-client/pull/574
2. refactor search config API (ref: `WidgetSerializer` and `/api/config/<gid>/qdjango/<pid>`)
3. deprecate setting `G3W_CLIENT_SEARCH_ENDPOINT = 'ows'` (until removal, the new default is: `'api'`)

Raruto committed:
move shorten condition at top (if-else)

Raruto committed:
ternary assignment

Raruto committed:
alias: `field['input']` → `input`

Raruto committed:
reduce nesting

Raruto committed:
keep a focus on JSON response structure

Raruto committed:
comments

Raruto committed:
spacing

Raruto committed:
client https://github.com/g3w-suite/g3w-client/commit/5bbe9ca5eecacaa0fbfcca1aa8800e00fcd94bc0

Raruto committed:
`widget_serializzer_data` → `w_data`

Raruto committed:
client https://github.com/g3w-suite/g3w-client/commit/c08802c848d9126eab0bf323d3b95469640a0772

Raruto commented on g3w-admin/qdjango/api/projects/serializers.py:
```diff
@@ -930,73 +927,63 @@ def __init__(self, instance=None, data=empty, **kwargs):
+                        'type': ({
+                            'autocompletebox': 'autocompletefield',
+                            'selectbox': 'selectfield'
+                        }).get(field.get('widgettype'), field['input'].get('type')),
```
@wlorenzetti the `numberfield` type should also be returned in here (currently not implement in v3.7 response):


## OK: "Volume greater than" appears as number field

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/d7840158-c5e6-4781-8f08-345e40471cc2)

## KO: "Volume greater than" appears as text field

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/39a8e6dc-2fa1-4485-ac7b-66baeb4fb6dc)

## Sample config: "Type and volume"

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/b3885ab6-b6a3-4020-8877-eb98ad09e780)

Raruto committed:
fix `search_1n`

Raruto committed:
TODO

Raruto commented on g3w-admin/base/settings/base.py:
```diff
@@ -291,6 +291,8 @@
 CLIENT_G3WSUITE_LOGO = 'g3wsuite_logo_h40.png'
 CLIENT_OWS_METHOD = 'GET'
+
+# TODO: set qgis 'api' as default search endpoint?
 G3W_CLIENT_SEARCH_ENDPOINT = 'ows' #or 'api' for to use api layer vector with FieldFilterBackend
```
@wlorenzetti some searches do not work with OWS + in the demo I see that is set as default:

```suggestion
G3W_CLIENT_SEARCH_ENDPOINT = 'api' # or 'ows' for to use api layer vector with FieldFilterBackend
```

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/3ec6e330-bcec-438b-bfba-68fbedf52727)

Raruto replied:
@wlorenzetti I think this option is deprecated, I get similar results on the stable branch (an empty select in that case):

> [`G3W_CLIENT_SEARCH_ENDPOINT`](https://g3w-suite.readthedocs.io/en/v3.7.x/settings.html#g3w-client-search-endpoint)
> 
> Search url endpoint for ‘searches calling’, default ows.
> - _ows:_ by wms search;
> - _api:_ by g3w-suite layer vector API.

## `G3W_CLIENT_SEARCH_ENDPOINT = 'ows'`

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/a1f0eec1-64c5-4232-ba2a-f9469e2a8c19)

## `G3W_CLIENT_SEARCH_ENDPOINT = 'api'`

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/ea91001f-a11c-4d84-9c80-e7bda3dffd6b)

wlorenzetti replied:
@raruto what type of search is it?

wlorenzetti replied:
Ahh ok I read just now is a _Value Relation_ is it right?

Raruto replied:
>  is a Value Relation is it right?

@wlorenzetti yes for the image above, but additionally (with the current branch) I think  you'll get: `No results found for this query/search` for almost all of them:

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/196e9534-1804-478e-8da8-b2545b653e2b)

wlorenzetti replied:
@volterra79 yes I get No results, but the response is correct, i.e. for _Type and sybtype_ search type: 
```
localhost:8006/ows/68/qdjango/336/?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetFeatureInfo&LAYERS=buildings_2f43dc1d_6725_42d2_a09b_dd446220104a&QUERY_LAYERS=buildings_2f43dc1d_6725_42d2_a09b_dd446220104a&INFO_FORMAT=application%2Fvnd.ogc.gml&FEATURE_COUNT=10000&CRS=EPSG%3A3857&I=0&J=0&FILTER=buildings_2f43dc1d_6725_42d2_a09b_dd446220104a%3A%22type%22+%3D+%27Administrative%27+AND+%22year%22+%3D+%272015%27&WITH_GEOMETRY=true
```

 I get a wms xml response :
```xml
<?xml version="1.0" encoding="UTF-8"?>
<wfs:FeatureCollection xmlns:qgs="http://qgis.org/gml" xmlns:ows="http://www.opengis.net/ows" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xlink="http://www.w3.org/1999/xlink" xsi:schemaLocation="http://www.opengis.net/wfs http://schemas.opengis.net/wfs/1.0.0/wfs.xsd http://qgis.org/gml" xmlns:wfs="http://www.opengis.net/wfs" xmlns:ogc="http://www.opengis.net/ogc" xmlns:gml="http://www.opengis.net/gml">
 <gml:featureMember>
  <qgs:buildings_2f43dc1d_6725_42d2_a09b_dd446220104a fid="buildings_2f43dc1d_6725_42d2_a09b_dd446220104a.48">
   <gml:boundedBy>
    <gml:Box srsName="EPSG:3857">
     <gml:coordinates cs="," ts=" ">1252098.43991029,5433342.30249331 1252122.43188242,5433364.29513444</gml:coordinates>
    </gml:Box>
   </gml:boundedBy>
   <qgs:geometry>
    <gml:MultiPolygon srsName="EPSG:3857">
     <gml:polygonMember>
      <gml:Polygon>
       <gml:outerBoundaryIs>
        <gml:LinearRing>
         <gml:coordinates cs="," ts=" ">1252098.43991029,5433348.8003191 1252102.97172725,5433364.29513444 1252122.43188242,5433352.53240365 1252119.49953027,5433342.30249331 1252098.43991029,5433348.8003191</gml:coordinates>
        </gml:LinearRing>
       </gml:outerBoundaryIs>
      </gml:Polygon>
     </gml:polygonMember>
    </gml:MultiPolygon>
   </qgs:geometry>
   <qgs:fid>48</qgs:fid>
   <qgs:name>aaaa</qgs:name>
   <qgs:address>Piazza Pietro Leopoldo</qgs:address>
   <qgs:year>2015</qgs:year>
   <qgs:photo>NULL</qgs:photo>
   <qgs:link>NULL</qgs:link>
   <qgs:form>NULL</qgs:form>
   <qgs:high>0</qgs:high>
   <qgs:volume>0</qgs:volume>
   <qgs:surface>153.5</qgs:surface>
   <qgs:architectural_barriers>()</qgs:architectural_barriers>
   <qgs:date_barriers>2024/03/29</qgs:date_barriers>
   <qgs:safety_exits>()</qgs:safety_exits>
   <qgs:date_exits>NULL</qgs:date_exits>
   <qgs:fire_system>()</qgs:fire_system>
   <qgs:date_fire>NULL</qgs:date_fire>
   <qgs:type>Administrative</qgs:type>
   <qgs:zone>B</qgs:zone>
   <qgs:attachment>false</qgs:attachment>
   <qgs:subtype>NULL</qgs:subtype>
   <qgs:user>NULL</qgs:user>
   <qgs:description>NULL</qgs:description>
  </qgs:buildings_2f43dc1d_6725_42d2_a09b_dd446220104a>
 </gml:featureMember>
</wfs:FeatureCollection>
```
but it is not show in g3w-client.

However I think the `G3W_CLIENT_SEARCH_ENDPOINT` option can be eliminated and default to 'ows' type.

Raruto replied:
> However I think the G3W_CLIENT_SEARCH_ENDPOINT option can be eliminated and default to 'ows' type.

@wlorenzetti perhaps you meant `'api'` 😉 

> I get No results, but the response is correct

For those who can debug them, firefox support for `application/vnd.ogc.gml` is quite terrible 😞:

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/30679ed1-c7cf-4a47-b754-f0fec42d133c)

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/8897a168-549e-4520-b9bd-37755bd437e4)

volterra79 committed:
:sparkles: Client
  https://github.com/g3w-suite/g3w-client/pull/574/commits/c8f16e9ff30497fddb2ea2831739223c1677ddd2
  https://github.com/g3w-suite/g3w-client-plugin-editing/pull/106/commits/fccb0e6ccdafea1421fc37585c0c0ac1c3d71ba2

Raruto committed:
https://github.com/g3w-suite/g3w-client/pull/574/commits/b83cdad495922a36ef411fb393de63dcf5a74669

Raruto committed:
https://github.com/g3w-suite/g3w-client/commit/f797285ff2ba7f39f41ec4245892d50c373fa439

Raruto committed:
Merge branch 'dev' into g3w-client_PR574

Raruto committed:
move `layerid` into `WidgetSerializer`

volterra79 committed:
:bug: Client
  https://github.com/g3w-suite/g3w-client-plugin-editing/pull/110/commits/d03671e2a650e581ce70aeed3d189b9e2f3ecdb4

Raruto committed:
Merge branch 'g3w-client_PR574' of https://github.com/g3w-suite/g3w-admin into g3w-client_PR574

Raruto committed:
https://github.com/g3w-suite/g3w-client/pull/574/commits/28435f298bb918e942f81bd5c3f3108f501d194e

Raruto committed:
deprecate `G3W_CLIENT_SEARCH_ENDPOINT`

Raruto committed:
https://github.com/g3w-suite/g3w-client/pull/574/commits/ea6e5acb587e1eaab97dcac5953874f36e97ba87

Raruto committed:
attempt to fix client test

Raruto committed:
Merge branch 'dev' into g3w-client_PR574

Raruto committed:
Merge branch 'dev' into g3w-client_PR574

volterra79 committed:
:bug: Client
  https://github.com/g3w-suite/g3w-client-plugin-editing/pull/110/commits/dc37654c7e92eb9a2c8aad4258fa1414a3b149e2

Raruto committed:
remove initial `values` in within search config

Raruto committed:
https://github.com/g3w-suite/g3w-client/pull/574

Raruto committed:
tests

Raruto committed:
`blanktext` → `description`

Raruto committed:
https://github.com/g3w-suite/g3w-client/pull/574/commits/1f3a8d9e69172abbc82c43e64758b663f50d8eee

volterra79 committed:
:bug: Client
 https://github.com/g3w-suite/g3w-client-plugin-editing/pull/110/commits/5211beee69465a26cfb2fede5a4e827cde3d333f

Raruto committed:
Merge branch 'dev' into g3w-client_PR574

Raruto committed:
https://github.com/g3w-suite/g3w-client/pull/574/commits/e108bcd22e02029e1d28cbb06323f425a1394061

Raruto committed:
https://github.com/g3w-suite/g3w-client/pull/574/commits/9618ec25757fc37fa834d1e591541694e4304ed7

Raruto commented:
@leolami https://github.com/g3w-suite/g3w-client/pull/574/commits/f99134b3cf75ca1c817065d4e63dd1756cf71017 takes care of empty query result for: `autompletefield` + `RelationReference`.

## Before

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/d115bfb4-e3c6-4022-866f-863d5c113770)


## After

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/1c95d06c-539a-406c-8f96-dbe3b8081762)


---

### Unsearchable layer(s)

I think some check is missing on the server side.

From what I see, some other layers shouldn't be searchable either (unless relationships could break that option?):


![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/ea2e6f42-09db-4148-bea9-c3fac1648c76)

Raruto committed:
Merge branch 'dev' into g3w-client_PR574

volterra79 committed:
:sparkles: Client
  https://github.com/g3w-suite/g3w-client-plugin-editing/pull/110/commits/037522210e81a9ff4086f4af58f06575e25da08c

volterra79 committed:
Merge branch 'g3w-client_PR574' of https://github.com/g3w-suite/g3w-admin into g3w-client_PR574

Raruto committed:
Merge branch 'dev' into g3w-client_PR574

Raruto commented:
@wlorenzetti still waiting for your review → merge 😉

wlorenzetti reviewed (approved):
I seems works fine, I just make some tests.
